### DISCUSSION TITLE
how to use MetaPhysicL::raw_value

### DISCUSSION THREAD

**Discussion by saturn00000** [idaholab/moose#21276] [Q&A General] [answered]
Hi MOOSEers,

When I try to  find **the direction angle for _grad_u[_qp]**, must I use "**raw_value**"?, 
`
ADReal theta= std::atan2(raw_value(_grad_u[_qp](1)), raw_value(_grad_u[_qp](0)));`

or  **_without_** "**raw_value**"
`
ADReal theta = std::atan2(_grad_u[_qp](1), _grad_u[_qp](0));`

 I appreciate any suggestions and help.Thank you.

Ref: https://github.com/idaholab/moose/discussions/20312

**Comment by GiudGiud** [comment, marked as answer]
Hello

raw_value is used to remove the derivatives, so to create a Real out of an ADReal. 

So it depends. If you are contributing to the residual with an ADKernel you want to use ADReal. If you are building an auxkernel or a postprocessor you can use Real. 

Guillaume

**Reply by saturn00000** [reply]
can we direct use **std::atan2(_grad_u[_qp](1), _grad_u[_qp](0))**?  or

**Comment by saturn00000** [comment]
Hello @GiudGiud,

if we use  **std::atan2(_grad_u[_qp](1), _grad_u[_qp](0))**  to the residual with an ADKernel , then we will have the **error**:

  ```
Linear solve did not converge due to DIVERGED_NANORINF iterations 0
Nonlinear solve did not converge due to DIVERGED_LINEAR_SOLVE iterations 0
 Solve Did NOT Converge!
  Finished Solving
```

**Reply by GiudGiud** [reply]
Hello

This can happen if the function derivative is not defined everywhere, especially if the function's derivative has a pole where its value is undefined.
However, atan's derivative is fairly well behaved so I'm not sure. 

While it is preferred to use ADReal everywhere and preserve all derivatives, if a term is difficult to handle you may remove its derivatives and have an imperfect Jacobian. You may also use an `if` condition to isolate the problematic area. 

Guillaume

**Reply by saturn00000** [reply]
Hello @GiudGiud 

thanks a lot for your reply!

I think the reason may be that that the (AD) derivative of  the function **atan2** is not defined in "MetaPhysicL".

P.S.  the (AD) derivative of  the function **atan** can be used in MOOSE, but not **atan2** ...
